src/sql/src/plan/query.rs

```diff
@@ -547,12 +548,130 @@ pub fn plan_insert_query(
         }
     }
 
-    Ok((
-        table.id(),
-        expr.map(map_exprs).project(project_key),
-        // TODO(asenac) return only the columns requested in the order requested
-        desc.clone(),
-    ))
+    Ok((table.id(), expr.map(map_exprs).project(project_key)))
+}
+
+pub fn plan_copy_from(
+    scx: &StatementContext,
+    table_name: UnresolvedObjectName,
+    columns: Vec<Ident>,
+) -> Result<(GlobalId, RelationDesc, Vec<usize>), anyhow::Error> {
+    let table = scx.resolve_item(table_name)?;
+
+    // Validate the target of the insert.
+    if table.item_type() != CatalogItemType::Table {
+        bail!(
+            "cannot insert into {} '{}'",
+            table.item_type(),
+            table.name()
+        );
+    }
+    let mut desc = table.desc()?.clone();
+    let _ = table
+        .table_details()
+        .expect("attempted to insert into non-table");
+
+    if table.id().is_system() {
+        bail!("cannot insert into system table '{}'", table.name());
+    }
+
+    let mut ordering = Vec::with_capacity(columns.len());
+
+    if columns.is_empty() {
+        ordering.extend(0..desc.arity());
+    } else {
+        let columns: Vec<_> = columns.into_iter().map(normalize::column_name).collect();
+        let column_by_name: HashMap<&ColumnName, (usize, &ColumnType)> = desc
+            .iter()
+            .filter_map(|(name, typ)| name.map(|n| (n, typ)))
+            .enumerate()
+            .map(|(idx, (name, typ))| (name, (idx, typ)))
+            .collect();
+
+        let mut names = Vec::with_capacity(columns.len());
+        let mut source_types = Vec::with_capacity(columns.len());
+
+        for c in &columns {
+            if let Some((idx, typ)) = column_by_name.get(c) {
+                ordering.push(*idx);
+                source_types.push((*typ).clone());
+                names.push(Some(c.clone()));
+            } else {
+                bail!(
+                    "column {} of relation {} does not exist",
+                    c.as_str().quoted(),
+                    table.name().to_string().quoted()
+                );
+            }
+        }
+        if let Some(dup) = columns.iter().duplicates().next() {
+            bail!("column {} specified more than once", dup.as_str().quoted());
+        }
+
+        desc = RelationDesc::new(RelationType::new(source_types), names);
+    };
+
+    Ok((table.id(), desc, ordering))
+}
+
+/// Builds a plan that adds the default values for the missing columns and re-orders
+/// the datums in the given rows to match the order in the target table.
+// TODO(asenac) This needs refactoring and a better name.
+pub fn plan_copy_from_2(
```

Comment:
This needs attention. In order to add the default values for the missing columns, this methods generates a plan with a constant operator containing the received rows, and then adds a map and a project adding the remaining columns.

COPY FROM STDIN is very similar to INSERT statement, so some refactoring seems possible, although COPY FROM STDIN requires a two-steps planning (at least with this solution).